### PR TITLE
[LLM Runtime] Add MX-Format (FP8_E5M2, FP8_E4M3, FP4_E2M1, NF4)

### PR DESCRIPTION
**Pull request by zhentaoyu**
## Type of Change

feature or bug fix or documentation or others
API changed or not: no

## Description

detail description 
JIRA ticket: xxx
rely on #769 

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Comment by zhentaoyu** [comment]
@a32543254 , @zhenwei-intel please have a look here. We may need to design the user interface.
1. like nv: `use_fp8=True` in `AutoModel.from_pretrained(xxx)`
2. image
3....

**Comment by airMeng** [comment]
> @a32543254 , @zhenwei-intel please have a look here. We may need to design the user interface

+ @PenghuiCheng @xin3he @changwangss FYI

**Comment by a32543254** [comment]
> @a32543254 , @zhenwei-intel please have a look here. We may need to design the user interface.
> 
> 1. like nv: `use_fp8=True` in `AutoModel.from_pretrained(xxx)`
> 2. image
> 3. .....



Let‘s have nf4/int4 performance comparison result first,
Haihao suggest If the performance loss is within 5%, we can accept it to get a better acc in return.

**Comment by a32543254** [comment]
> @a32543254 , @zhenwei-intel please have a look here. We may need to design the user interface.
> 
> 1. like nv: `use_fp8=True` in `AutoModel.from_pretrained(xxx)`
> 2. image
> 3. .....

Yes, our API should get close to transformers like, we could use same args what transformers already have.

**Comment by zhentaoyu** [comment]
`FP4` & `NF4` test
env: SPR (m6), 1 ins / 55 core, `jblas` pr commit: https://github.com/intel/intel-extension-for-transformers/pull/769/commits/63d5c9d7eda8d92c8a21cc444f6d329408d0db94

first token logits compare:
```python
def cmpData(numa, numb):
    totalErr = ((numa - numb)**2).sum()
    totalNum = (numa**2).sum()
    diff2 = np.sqrt(totalErr / totalNum)

    cos = np.dot(numa, numb) / (np.linalg.norm(numa) * np.linalg.norm(numb))
    return {"diff2": diff2, "cos": cos}
```
int4 / fp32: {'diff2': 0.11176497, 'cos': 0.9937631}
fp4 / fp32: {'diff2': 0.1053754, 'cos': 0.9944502} 
nf4 / fp32 {'diff2': 0.07750923, 'cos': 0.9970596}
**NF4 seems better**

perf:
|  weight_dtype  |  first token (avg, ms)  |  next token (avg, ms) |
|  -------       | ------                  | -----                 |
|   fp32         |    56.78  |  131.38  |       
|   int4         |    3.72   |    32.39 |
|   fp4          |     3.77  |    34.85 |
|   nf4          |     3.96    |  33.67 |

**Comment by a32543254** [comment]
> `FP4` & `NF4` test env: SPR (m6), 1 ins / 55 core, `jblas` pr commit: [63d5c9d](https://github.com/intel/intel-extension-for-transformers/commit/63d5c9d7eda8d92c8a21cc444f6d329408d0db94)
> 
> first token logits compare:
> 
> ```python
> def cmpData(numa, numb):
>     totalErr = ((numa - numb)**2).sum()
>     totalNum = (numa**2).sum()
>     diff2 = np.sqrt(totalErr / totalNum)
> 
>     cos = np.dot(numa, numb) / (np.linalg.norm(numa) * np.linalg.norm(numb))
>     return {"diff2": diff2, "cos": cos}
> ```
> 
> int4 / fp32: {'diff2': 0.11176497, 'cos': 0.9937631} fp4 / fp32: {'diff2': 0.1053754, 'cos': 0.9944502} nf4 / fp32 {'diff2': 0.07750923, 'cos': 0.9970596} **NF4 seems better**
> 
> perf: `compute_dtye=bf16, group_size=32, alg=sym, scale_dtype=bf16`
> 
> weight_dtype	first token (avg, ms)	next token (avg, ms)
> fp32	56.78	131.38
> int4	3.72	32.39
> fp4	3.77	34.85
> nf4	3.96	33.67
> `compute_dtye=fp32, group_size=32, alg=sym, scale_dtype=fp32`
> 
> weight_dtype	first token (avg, ms)	next token (avg, ms)
> int4	6.91	27.77
> fp4	6.87	28.43
> nf4	6.46	28.26

Based on following data, we could control performance loss within 5% and get acc gain,
So please set nf4 as default for load_in_4bit.
Refer to change here https://github.com/intel/intel-extension-for-transformers/blob/cf9d3803512ff9603a453d753138d6a0a8df85d3/intel_extension_for_transformers/transformers/modeling/modeling_auto.py#L124

**Comment by zhentaoyu** [comment]
After discussed with @a32543254 and @zhenwei-intel , llm runtime (or neural speed) will continue using `WeightOnlyConfig` to interact with `ITREX` frontend. Other origin transformers repo's quant config will update the final params of `WeightOnlyConfig`.

> > @a32543254 , @zhenwei-intel please have a look here. We may need to design the user interface
> 
> * @PenghuiCheng @xin3he @changwangss FYI

**Review by zhenwei-intel — APPROVED** [review]

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by airMeng** [comment]
can you update CI, at least we need to maintain llama2
https://github.com/intel/intel-extension-for-transformers/blob/43e30bcabd766620a337f8aaa7d5bbfa6411c246/.github/workflows/script/models/cpp_graph_inference.sh#L23-L28

@intellinjun please verify QWen MX.  When will we support QWen in CI?

**Comment by VincyZhang** [comment]
ready for merge?

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/206/

**Comment by intellinjun** [comment]
> can you update MX validation in CI, at least we need to maintain llama2
> 
> https://github.com/intel/intel-extension-for-transformers/blob/43e30bcabd766620a337f8aaa7d5bbfa6411c246/.github/workflows/script/models/cpp_graph_inference.sh#L23-L28
> 
> @intellinjun please verify QWen MX. When will we support QWen in CI?

Maybe we should update in [here](https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/blob/main/nlp-toolkit/scripts/cpp_graph_inference.sh#L115).

**Comment by zhentaoyu** [comment]
> > can you update MX validation in CI, at least we need to maintain llama2
> > https://github.com/intel/intel-extension-for-transformers/blob/43e30bcabd766620a337f8aaa7d5bbfa6411c246/.github/workflows/script/models/cpp_graph_inference.sh#L23-L28
> > 
> > @intellinjun please verify QWen MX. When will we support QWen in CI?
> 
> Maybe we should update in [here](https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/blob/main/nlp-toolkit/scripts/cpp_graph_inference.sh#L115).

see here, wenxin already make a pr
https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/commit/ed4f0f9904ae78a48807795430ce2c35aafbb6be

**Comment by zhentaoyu** [comment]
> > > can you update MX validation in CI, at least we need to maintain llama2
> > > https://github.com/intel/intel-extension-for-transformers/blob/43e30bcabd766620a337f8aaa7d5bbfa6411c246/.github/workflows/script/models/cpp_graph_inference.sh#L23-L28
> > > 
> > > @intellinjun please verify QWen MX. When will we support QWen in CI?
> > 
> > 
> > Maybe we should update in [here](https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/blob/main/nlp-toolkit/scripts/cpp_graph_inference.sh#L115).
> 
> see here, wenxin already make a pr [intel-innersource/frameworks.ai.lpot.lpot-validation@ed4f0f9](https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/commit/ed4f0f9904ae78a48807795430ce2c35aafbb6be)

@VincyZhang ,`fp8_e5m2` still has a issue here which will be fixed in https://github.com/intel/intel-extension-for-transformers/pull/915. So the ext test report of `fp8_e5m2` may have bad outputs in this pr

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/209/

**Review by airMeng — APPROVED** [review]

**Comment by VincyZhang** [comment]
extension test pass, please check whether performance achieve expectation. @zhentaoyu @zhewang1-intc @luoyu-intel @airMeng
